### PR TITLE
Open source prescriptor training using PyTorch implementation of NSGA-II

### PR DESCRIPTION
**Pull request by danyoungday**
Restructured project a bit to enable multiple types of prescriptor.

The abstract Prescriptor class requires prescribe_land_use and predict_metrics to be implemented as that's all that's really needed from our experiments - the ability to go from context to context + actions and the ability to see what these context + actions result in from the predictor.

The old ESP prescriptor is relatively unchanged and just tweaked a little to implement these functions.

The new TorchPrescriptor is trained using NSGA-II and also implements these functions.

The prescriptor experiments from the paper have been modified to take in a Prescriptor so we can rerun them using the TorchPrescriptor and compare the results.

This also enables us to implement different types of prescriptors. For example, for future work I will probably change the heuristics to be implementations of the Prescriptor class rather than manually called functions in the experiments notebook.

### CODE REVIEW

File: use_cases/eluc/.gitignore [hunk 1]
```diff
@@ -4,9 +4,11 @@ predictors/*/trained_models/
 # Ignores predictor significance results
 experiments/predictor_significance
 
+experiments/figures
+
 # Ignores trained prescriptors and seeds
-prescriptors/trained_prescriptors
-prescriptors/seeds
+prescriptors/*/trained_prescriptors
+prescriptors/*/seeds
```
Comment (danyoungday) [inline use_cases/eluc/.gitignore:11]: New project structure with abstract prescriptor class means we need this to look the same as the predictors

File: use_cases/eluc/data/eluc_data.py [hunk 2]
```diff
@@ -27,7 +27,13 @@ def encode_as_df(self, df: pd.DataFrame) -> pd.DataFrame:
...
         for col in new_df.columns:
             if col in self.fields:
-                new_df[col] = (new_df[col] - self.fields[col]["range"][0]) / (self.fields[col]["range"][1] - self.fields[col]["range"][0])
+                min_val = self.fields[col]["range"][0]
+                max_val = self.fields[col]["range"][1]
+                # If min and max are the same, just set value to 0
+                if min_val == max_val:
+                    new_df[col] = 0
+                else:
+                    new_df[col] = (new_df[col] - self.fields[col]["range"][0]) / (self.fields[col]["range"][1] - self.fields[col]["range"][0])
```
Comment (danyoungday) [inline use_cases/eluc/data/eluc_data.py:36]: Fixed an edge case where all values in a column are the same. We just scale this to 0, however maybe it's better to scale to 0.5?

File: use_cases/eluc/data/torch_data.py
Comment (danyoungday) [inline use_cases/eluc/data/torch_data.py:1]: Moved custom dataset used to train NeuralNetPredictor into its own class so the TorchPrescriptor can use it too

File: use_cases/eluc/prescriptors/esp/unileaf_prescriptor.py
Comment (danyoungday) [inline use_cases/eluc/prescriptors/esp/unileaf_prescriptor.py:1]: Moved everything from old prescriptor.py into esp/unileaf_prescriptor.py

File: use_cases/eluc/prescriptors/esp/unileaf_prescriptor.py [hunk 3]
```diff
@@ -0,0 +1,305 @@
...
+        context_actions_df = context_actions_df.set_index(context_df.index)
+
+        return context_actions_df
+    
+    def predict_metrics(self, context_actions_df: pd.DataFrame) -> tuple:
+        eluc_df = self.predict_eluc(context_actions_df)
+        change_df = self.compute_percent_changed(context_actions_df)
+
+        return eluc_df, change_df
+
```
Comment (danyoungday) [inline use_cases/eluc/prescriptors/esp/unileaf_prescriptor.py:304]: Added these 2 bottom functions to implement the Prescriptor.py abstract class. They are just simple calls of already made functions.

File: use_cases/eluc/prescriptors/nsga2/candidate.py
Comment (danyoungday) [inline use_cases/eluc/prescriptors/nsga2/candidate.py:1]: Simple PyTorch prescriptor candidates. They're a 1 hidden layer fixed-architecture feed forward neural network.

File: use_cases/eluc/prescriptors/nsga2/candidate.py [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+            torch.nn.Linear(hidden_size, out_size))
+        
+        self.device = device
+        self.model.to(device)
+        self.model.eval()
+        
+        # Orthogonal initialization
+        for layer in self.model:
+            if isinstance(layer, torch.nn.Linear):
+                torch.nn.init.orthogonal_(layer.weight)
```
Comment (danyoungday) [inline use_cases/eluc/prescriptors/nsga2/candidate.py:31]: Orthogonal intialization

File: use_cases/eluc/prescriptors/nsga2/candidate.py [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+                layer.bias.data.fill_(0.01)
+
+        # To keep track of metrics
+        self.metrics = None
+        self.rank = None
+        self.distance = -1
```
Comment (danyoungday) [inline use_cases/eluc/prescriptors/nsga2/candidate.py:37]: We keep track of our own metrics rather than having them kept in some external dict in the NSGA-II algorithm code.

File: use_cases/eluc/prescriptors/nsga2/create_seeds.py
Comment (danyoungday) [inline use_cases/eluc/prescriptors/nsga2/create_seeds.py:1]: Same code as the ESP seed training just implemented in PyTorch

File: use_cases/eluc/prescriptors/nsga2/nsga2_utils.py
Comment (danyoungday) [inline use_cases/eluc/prescriptors/nsga2/nsga2_utils.py:1]: NSGA-II helper functions generated by Elliot's ChatGPT. Hand tested them a little and it seems like they work but more thorough testing is probably required down the line.

File: use_cases/eluc/prescriptors/nsga2/nsga2_utils.py [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+            distances[i] += (front[i+1].metrics[m] - front[i-1].metrics[m]) / (obj_max - obj_min)
+
+    return distances
+
+def dominates(candidate1: Candidate, candidate2: Candidate):
+    """
+    Determine if one individual dominates another.
+    """
+    for obj1, obj2 in zip(candidate1.metrics, candidate2.metrics):
+        if obj1 > obj2:
```
Comment (danyoungday) [inline use_cases/eluc/prescriptors/nsga2/nsga2_utils.py:82]: Currently dominates means less than so all objectives are being minimized. At the moment to maximize an outcome we have to negative its values but for now this works fine.

File: use_cases/eluc/prescriptors/nsga2/torch_prescriptor.py [hunk 7]
```diff
@@ -0,0 +1,264 @@
...
+        # Sum the positive diffs
+        percent_changed = context_actions_df[context_actions_df[constants.DIFF_LAND_USE_COLS] > 0][constants.DIFF_LAND_USE_COLS].sum(axis=1)
+        # Divide by sum of used land
+        total_land = context_actions_df[constants.LAND_USE_COLS].sum(axis=1)
+        total_land = total_land.replace(0, 1) # Avoid division by 0
+        percent_changed = percent_changed / total_land
+        change_df = pd.DataFrame(percent_changed, columns=["change"])
+        return change_df
+    
+    def _prescribe(self, candidate: Candidate, context_df=None) -> pd.DataFrame:
```
Comment (danyoungday) [inline use_cases/eluc/prescriptors/nsga2/torch_prescriptor.py:93]: Flow of prescription is: context df -> context tensor -> goes through model -> reco tensor -> reco df -> modify context df with reco df to get context + actions df

File: use_cases/eluc/prescriptors/nsga2/torch_prescriptor.py [hunk 8]
```diff
@@ -0,0 +1,264 @@
...
+        self.n_generations = n_generations
+        self.p_mutation = p_mutation
+        self.seed_dir=seed_dir
+
+        self.eval_df = eval_df
+        self.encoded_eval_df = encoder.encode_as_df(eval_df)
+        self.encoder = encoder
+        context_ds = TorchDataset(self.encoded_eval_df[constants.CAO_MAPPING["context"]].to_numpy(),
+                                  np.zeros((len(self.encoded_eval_df), len(constants.RECO_COLS))))
+        self.context_dl = torch.utils.data.DataLoader(context_ds, batch_size=batch_size, shuffle=False)
```
Comment (danyoungday) [inline use_cases/eluc/prescriptors/nsga2/torch_prescriptor.py:48]: We cache the evaluation dataset so that we don't have to re-convert from pandas to torch over and over during training. Then during inference we can pass our own dataframe in.

File: use_cases/eluc/prescriptors/nsga2/torch_prescriptor.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+        """
+        Same implementation as in ESP.
+        Takes two random parents and compares their indices since this is a measure of their performance.
+        Note: It is possible for this function to select the same parent twice.
+        """
+        idx1 = min(random.choices(range(len(sorted_parents)), k=2))
+        idx2 = min(random.choices(range(len(sorted_parents)), k=2))
+        return sorted_parents[idx1], sorted_parents[idx2]
+
+    def _make_new_pop(self, parents: list[Candidate], pop_size: int, gen:int) -> list[Candidate]:
```
Comment (danyoungday) [inline use_cases/eluc/prescriptors/nsga2/torch_prescriptor.py:175]: No elites are added to the new population because NSGA-II already does that inherently

File: use_cases/eluc/prescriptors/prescriptor.py
Comment (danyoungday) [inline use_cases/eluc/prescriptors/prescriptor.py:1]: This diff looks really weird because it originally housed the UnileafPrescriptor class but now holds the abstract Prescriptor class.

File: use_cases/eluc/tests/test_nsga2.py
Comment (danyoungday) [inline use_cases/eluc/tests/test_nsga2.py:1]: Some basic unit tests. A LOT more are going to be needed to fully verify everything is working as intended, although I fixed so many bugs already hopefully everything works fine.

File: use_cases/eluc/data/eluc_data.py (hunk 2 above)
Comment (ofrancon) [reply use_cases/eluc/data/eluc_data.py:36]: I think 0 should be fine - the scaler will map it to the min value, which is also the max value.
That being said, if all the values in the column are the same, the column is probably not useful...

File: use_cases/eluc/prescriptors/esp/unileaf_prescriptor.py (hunk 3 above)
Comment (ofrancon) [reply use_cases/eluc/prescriptors/esp/unileaf_prescriptor.py:304]: Maybe we should make theme more generic in a later iteration, e.g. `prescribe(...)` and `compute_metrics(...)`.

File: use_cases/eluc/prescriptors/nsga2/nsga2_utils.py (hunk 6 above)
Comment (ofrancon) [reply use_cases/eluc/prescriptors/nsga2/nsga2_utils.py:82]: That's fine for the land use change domain but we'll have to be careful for future domains.

Comment (ofrancon) [reply use_cases/eluc/prescriptors/nsga2/nsga2_utils.py:82]: This kind of mistake is usually obvious to plot on the pareto front: we should quickly see we're minimizing instead of maximizing

File: use_cases/eluc/prescriptors/nsga2/torch_prescriptor.py (hunk 8 above)
Comment (ofrancon) [reply use_cases/eluc/prescriptors/nsga2/torch_prescriptor.py:48]: Worth a comment in the code directly.

File: use_cases/eluc/prescriptors/nsga2/torch_prescriptor.py (hunk 7 above)
Comment (ofrancon) [reply use_cases/eluc/prescriptors/nsga2/torch_prescriptor.py:93]: Could also be a useful comment in the code or docstring

File: use_cases/eluc/prescriptors/nsga2/torch_prescriptor.py [hunk 10]
```diff
@@ -0,0 +1,265 @@
...
+            if len(parents) + len(front) > n_parents:  # If adding this front exceeds num_parents
+                front = sorted(front, key=lambda c: c.distance, reverse=True)
+                parents += front[:n_parents - len(parents)]
+                break
+            parents += front
+        return parents
+    
+    def _tournament_selection(self, sorted_parents: list[Candidate]) -> tuple[Candidate, Candidate]:
+        """
+        Same implementation as in ESP.
```
Comment (ofrancon) [inline use_cases/eluc/prescriptors/nsga2/torch_prescriptor.py:167]: ESP is not open source, at least for now. I think you can remove references to it.

**Review by ofrancon — APPROVED** [review]
lgtm. @EKMeyerson feel free to review and iterate.

File: use_cases/eluc/prescriptors/nsga2/candidate.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+    def mutate(self, p_mutation: float):
+        """
+        Randomly mutates each weight with probability p_mutation with gaussian noise mu=0, sigma=0.1
+        """
+        for layer in self.model:
+            if isinstance(layer, torch.nn.Linear):
+                mask = torch.rand(size=layer.weight.shape) < p_mutation
+                layer.weight.data += mask * torch.randn(size=layer.weight.shape) * 0.1
+                mask = torch.rand(size=layer.bias.shape) < p_mutation
+                layer.bias.data += mask * torch.randn(size=layer.bias.shape) * 0.1
```
Comment (EKMeyerson) [inline use_cases/eluc/prescriptors/nsga2/candidate.py:80]: These `0.1` above could be a param to mutate, e.g., `mutation_std=0.1`

File: use_cases/eluc/prescriptors/nsga2/nsga2_utils.py
Comment (EKMeyerson) [reply use_cases/eluc/prescriptors/nsga2/nsga2_utils.py:1]: they look (aestethically) like a python version of pseudocode copied directly from the original nsga-ii paper. most important thing is they work :)
